### DISCUSSION TITLE
Controller Reverb G2 in Unity

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/MixedRealityToolkit-Unity#9071] [Q&A] [answered]
Hi there,
i am not new to unity but to vr and mrtk. I don't know if this is the right place for my question, but so far no one else has been able to help me.
I use the Reverb G2. At Unity I installed version 2020.2 and the latest version of mrtk and openxr plug-in 0.1.1.
As a controller I only get the debug arrow. I've now spent hours getting the reverb g2 controller into unity. I can not do it.
With steamvr none of this was a problem, but I would like to work without steamvr.
Thanks for any help.

**Comment by keveleigh** [comment, marked as answer]
Are you looking for the controller model of the physical controller itself? That's currently not exposed through the OpenXR plug-in but is on my backlog for the next few months. I can keep this thread updated with the version it ships in.

**Reply by ghost** [reply]
Hi there,
thanks for the answer. Yes, I mean the model of the g2 controller. Just as I see it in Windows Mixed Reality home. I suspected that the MRTK would find and use the controller model in the system and that I just didn't configure it correctly.

It's really a shame and I am astonished that it doesn't work. When I install steamvr and look at the simple sample, I immediately have the controller model fully functional, perfect. The button binding for the new G2 controller also works immediately.

I wanted to do something without SteamVR but will stick with it for now.

Happy new and healthy year

**Reply by polar-kev** [reply]
Hey @SaschaFFM, controller models will be available at runtime in MRTK 2.7 thanks to this PR, https://github.com/microsoft/MixedRealityToolkit-Unity/pull/9550

You can see it in the latest MRTK 2.7 preview package available through the MR Feature Tool.

**Reply by VRNerd96** [reply]
Hey @polar-kev  I can not find the prefabs for the controller. In which folder should I actually find them? Greetings

**Reply by RogPodge** [reply]
The models are loaded directly from the platforms firmware itself. There aren't any hp reverb models inside the project itself. If you have the backend configured correctly, the models will load appropriately during runtime.

**Comment by keveleigh** [comment]
@SaschaFFM Controller render models through OpenXR is supported in MRTK as of 2.7.0-preview.4 and will be supported in the next Mixed Reality OpenXR plugin release, version 0.9.4.

**Comment by sebstax7** [comment]
@SaschaFFM Could you please share your MRTK input and input data providers set up?

On my end i'm stuck at finding what to fill in MRTK inspector fields in "INPUT" and "INPUT DATA PROVIDER"?
I get a warning message  "this profile is not supported..." see my screenshot

That would be perfect if we can get a tutorial such as that one : 

https://docs.microsoft.com/en-us/windows/mixed-reality/develop/unity/tutorials/mr-learning-base-03?tabs=winxr

but for windows mixed reality.

I installed MRTK 2.6.1 and MR Input package along with OPENXR 1.1.1 plug in on Unity 2020.3.7f1LTS

![bug](https://user-images.githubusercontent.com/83447881/118017586-16bddb80-b357-11eb-8b57-93d8cfc0c36f.PNG)

**Reply by david-c-kline** [reply]
Hi @sebstax7. The OpenXR plugin that @keveleigh mentions is a Microsoft provided addon for HoloLens and Windows Mixed Reality specific functionality. It has not yet released (should be fairly soon) and is installable using the [Mixed Reality Feature Tool](https://aka.ms/mrfeaturetool).

![image](https://user-images.githubusercontent.com/13281406/118053106-c2a3fe80-b338-11eb-8e02-0f9600eb0f5b.png)

Once version 0.94 is released and imported, using MRTK v2.7.0-preview.4 or later should enable the G2 controller models.

![image](https://user-images.githubusercontent.com/13281406/118053238-01d24f80-b339-11eb-9269-995b8d8c45a7.png)

**Reply by david-c-kline** [reply]
@RogPodge, do you have any other input on this issue?

**Reply by RogPodge** [reply]
Nope, the once the newest packages are installed MRTK should pick up and render the controller models automatically. You shouldn't need to make modifications to our default set of profiles.

**Reply by sebstax7** [reply]
Hi @davidkline-ms @RogPodge @keveleigh @polar-kev 

 I installed MRTK 2.7 + Open XR 0.9.5 + Inputs using MixedRealityFeatureTool but It still does not work.
HP Reverb G2 controllers does not show up when I press play. 

Can you please explain how to set up the hierarchy in unity for MRTK and WMR controllers? With steamVR the process is to add two 3D models, add a script, press play, and it works fine.

Is there a WMR demo scene available? Can your Q&A team share their PC and software configuration?

Many thanks for your support



![features](https://user-images.githubusercontent.com/83447881/120505090-f4066c00-c3c4-11eb-9117-5226f0458a60.PNG)
![Capture](https://user-images.githubusercontent.com/83447881/120505100-f5d02f80-c3c4-11eb-9769-9388950f6f7a.PNG)
![Capturer](https://user-images.githubusercontent.com/83447881/120505251-17c9b200-c3c5-11eb-8e86-c280c42b7e09.PNG)

**Reply by keveleigh** [reply]
@sebstax7

1. Can you share the initial exception in your console? Usually the IsEnabled logs are preceded by some other initial exception
2. Can you double-check your Project Settings -> XR Plugin Management -> OpenXR feature sets? You'll need Motion Controller Model enabled for the platforms you're using (Standalone for the Editor).

**Reply by sebstax7** [reply]
@keveleigh  thanks

Problem solved, I confirm controllers works fine.
My mistake was I had checked "Windows Mixed Reality" in XR PLUG in management instead of "OPEN XR" and then, checked Windows Mixed reality which only appear once "OPEN XR" is checked.

**Comment by MiPa12** [comment]
Hi, I installed MRTK 2.7 + Open XR 0.9.5 + Inputs using MixedRealityFeatureTool + Unity Version 2020.3.11. I also enabled the HP Reverb G2 Controller Profile in the OpenXR Settings. Everything is working fine except the rendering of the HP Reverb G2 controllers. In my case it is just rendering the hand model (RiggedHandLeft, RiggedHandRight). Are there HP Reverb G2 Controller prefabs or how can I render the controllers in unity? Or do I have the wrong settings? Thanks.

![Screenshot (22)](https://user-images.githubusercontent.com/61323363/121746012-82859680-cb05-11eb-99a1-1148ded5c91b.png)
![Screenshot (24)](https://user-images.githubusercontent.com/61323363/121746079-9a5d1a80-cb05-11eb-8f00-da43fe4ec572.png)

**Reply by RogPodge** [reply]
Tagging @keveleigh since it's related to OpenXR

**Reply by RogPodge** [reply]
The models are loaded directly from the platforms firmware itself. There aren't any hp reverb models inside the project itself. If you have the backend configured correctly, the models will load appropriately during runtime.
